### PR TITLE
Adding a `Matches` method for `spy`'d methods

### PR DESCRIPTION
**Pull request by Nysosis**
This allows you to use a delegate method to check an argument was valid or not, as opposed to only being able to state what 'type' it was

Created abstract class `ArgumentMatcher` which most things use for logic now, instead of the `TypeMatcher`. `ArgumentMatcher` has method to test an object, and method to return a string representation of the matching (for failure messages)

# Description

# Checklist

- [x] I am an awesome developer and proud of my code
- [x] I added / updated / removed relevant unit or integration tests to prove my change works
- [x] I ran all tests using ```npm test``` to make sure everything else still works
- [x] I ran ```npm run review``` to ensure the code adheres to the repository standards

# Additional Information

### CODE REVIEW

**Comment by Nysosis** [comment]
@jamesrichford I've done that now (locally), however as I essentially did a commit reversal on https://github.com/alsatian-test/alsatian/commit/3422c1c032226acefaecb00185cc06d48caa7015 and started afresh, is it worth closing this pull request and starting a new one which is only the one commit, instead of the three?

**Comment by jamesrichford** [comment]
@Nysosis whichever is easier for you, I'm not too precious about commits as long as they say what they're doing then that's fine with me :)

File: core/spying/type-matcher.ts [hunk 1]
```diff
@@ -1,30 +1,124 @@
...
-         return value instanceof this._type;
+
+      if (typeof first === "object") {
+         return this.matchesObjectLiteral(first);
       }
+
+      throw new Error("Invalid arguments");
+   }
+
+   private matchesKeyAndValue(key: string, value: any): this {
```
Comment (jamesrichford) [inline core/spying/type-matcher.ts]: If we could add underscores before these private functions that'd be perfect! :)

**Comment by jamesrichford** [comment]
@Nysosis awesome!! Looks great :) your other changes so far should have been released yesterday in 1.2.0 by the way

**Comment by Nysosis** [comment]
Thanks!

Sorry took so long to make that private methods tweak, was away over the weekend.

**Comment by jamesrichford** [comment]
@Nysosis no problem at all, hope you had a great weekend!

Thanks again for another great PR 👍

**Pull request merged**